### PR TITLE
Make reusable base class for auxiliary indices

### PR DESCRIPTION
**Pull request by jimpo**
This refactors most of the logic in TxIndex into a reusable base class for other indices. There are two commits moving code between files, which may be be more easily reviewed using `git diff --color-moved` (https://blog.github.com/2018-04-05-git-217-released/).

The motivation for this is to support BIP 157 by indexing block filters.

---
This change is [Reviewable](https://reviewable.io/reviews/bitcoin/bitcoin/13243)

### CODE REVIEW

File: src/txdb.h [hunk 1]
```diff
@@ -125,6 +125,19 @@ class CBlockTreeDB : public CDBWrapper
...
 };
 
+class BaseIndexDB : public CDBWrapper
+{
+public:
+    BaseIndexDB(const fs::path& path, size_t n_cache_size,
+                bool f_memory = false, bool f_wipe = false, bool f_obfuscate = false);
+
+    /// Read block locator of the chain that the txindex is in sync with.
+    bool ReadBestBlock(CBlockLocator& locator) const;
```
Comment (promag) [inline src/txdb.h]: Remove `ReadBestBlock` and `WriteBestBlock` from `TxIndexDB` definition.

File: src/txdb.h [hunk 2]
```diff
@@ -125,6 +125,19 @@ class CBlockTreeDB : public CDBWrapper
     bool LoadBlockIndexGuts(const Consensus::Params& consensusParams, std::function<CBlockIndex*(const uint256&)> insertBlockIndex);
 };
 
+class BaseIndexDB : public CDBWrapper
```
Comment (promag) [inline src/txdb.h]: A comment explaining this base class would be nice.

File: src/Makefile.am [hunk 3]
```diff
@@ -103,6 +103,7 @@ BITCOIN_CORE_H = \
   fs.h \
   httprpc.h \
   httpserver.h \
+  index/base.h \
```
Comment (promag) [inline src/Makefile.am]: How about `index/baseindex.h`?

File: src/init.cpp [hunk 4]
```diff
@@ -1587,8 +1587,7 @@ bool AppInitMain()
 
     // ********************************************************* Step 8: start indexers
     if (gArgs.GetBoolArg("-txindex", DEFAULT_TXINDEX)) {
-        auto txindex_db = MakeUnique<TxIndexDB>(nTxIndexCache, false, fReindex);
-        g_txindex = MakeUnique<TxIndex>(std::move(txindex_db));
+        g_txindex = MakeUnique<TxIndex>(nTxIndexCache, false, fReindex);
```
Comment (promag) [inline src/init.cpp]: Why is this being changed?

Comment (jimpo) [reply src/init.cpp]: I think it makes sense for the DB to be encapsulated in the index interface, even if it is implemented separately internally.

File: src/txdb.h (hunk 2 above)
Comment (jimpo) [reply src/txdb.h]: Given that it gets moved to BaseIndex for better context in "index: Move index DBs into index/ directory.", would you still find the comment helpful?

File: src/txdb.h (hunk 1 above)
Comment (jimpo) [reply src/txdb.h]: Good catch, thanks.

File: src/Makefile.am (hunk 3 above)
Comment (promag) [reply src/Makefile.am]: Well I just think base too is vague. Also, there is `txindex.*`.

File: src/init.cpp (hunk 4 above)
Comment (promag) [reply src/init.cpp]: Agree with being encapsulated. It's just a unrelated change AFAIU, but a nice to include.

File: src/txdb.h (hunk 2 above)
Comment (promag) [reply src/txdb.h]: Right, ignore then.

File: src/index/base.cpp [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+    if (!GetDB().WriteBestBlock(locator)) {
+        error("%s: Failed to write locator to disk", __func__);
+    }
+}
+
+bool BaseIndex::BlockUntilSyncedToCurrentChain()
+{
+    AssertLockNotHeld(cs_main);
+
+    if (!m_synced) {
```
Comment (jamesob) [inline src/index/base.cpp]: I find this name a bit confusing ("if we're not synced and this method is called `BlockUntilSynced...`, why are we returning?"). Would a rename to `m_initial_sync_complete` or something along those lines make sense or does that seem too wordy?

**Review by jamesob — COMMENTED** [review]
Changes look really nice so far. Excited for how straightforward this will make adding more optional indexes. I'll start manual testing in the next few days.

File: src/index/base.cpp (hunk 5 above)
Comment (jimpo) [reply src/index/base.cpp]: Yeah, I see your point. I'm fine changing the variable name to be more descriptive `m_initial_sync_complete` or `m_sync_complete` both make sense to me (though the latter may not alleviate the confusion).

**Comment by promag** [comment]
utACK da17ca6. Commit by commit refactors LGTM.

**Review by jamesob — APPROVED** [review]
Tested ACK https://github.com/bitcoin/bitcoin/pull/13243/commits/da17ca6cd6539a6f8dc8ef0d76bb670e20feac67

Did a manual test of txindex behavior. Take or leave my previous comment about member attribute naming. Great change.

### Manual test plan


- [x] run -reindex (testnet)
- [x] set txindex=0, ensure that getrawtransaction doesn't work (testnet)
  ```
   $ which trpc
   trpc: aliased to ~/src/bitcoin/src/bitcoin-cli -testnet -rpcuser=foo -rpcpassword=bar

   $ trpc getrawtransaction $(trpc getblock `trpc getblockhash 500123` 1 | jq -r ".tx[3]")

   error code: -5
   error message:
   No such mempool transaction. Use -txindex to enable blockchain transaction queries. Use gettransaction for wallet transactions.
   ```
- [x] set txindex=1, ensure that getrawtransaction works again without a full reindex (testnet)
   ```
   [debug.log]
   2018-05-25T17:55:00Z Syncing txindex with block chain from height 153579
   2018-05-25T17:55:31Z Syncing txindex with block chain from height 208298
   2018-05-25T17:55:00Z Syncing txindex with block chain from height 153579
   2018-05-25T17:55:31Z Syncing txindex with block chain from height 208298
   2018-05-25T17:56:02Z Syncing txindex with block chain from height 276717
   2018-05-25T17:56:33Z Syncing txindex with block chain from height 506327
   2018-05-25T17:57:04Z Syncing txindex with block chain from height 628106
   2018-05-25T17:57:35Z Syncing txindex with block chain from height 894194
   2018-05-25T17:58:06Z Syncing txindex with block chain from height 1060431
   2018-05-25T17:58:37Z Syncing txindex with block chain from height 1162648
   2018-05-25T17:59:08Z Syncing txindex with block chain from height 1226430
   2018-05-25T17:59:39Z Syncing txindex with block chain from height 1285766
   2018-05-25T17:59:57Z txindex is enabled at height 1315371
   2018-05-25T17:59:57Z txindex thread exit

   $ trpc getrawtransaction $(trpc getblock `trpc getblockhash 500123` 1 | jq -r ".tx[3]")

   0100000001272bba5a7cb395ed0aad4360f8ecd2f3f6aa0089e695a40dc44f0d3d13d209e9010000006b483045022100c88339fe420b4d83a1a42bf8dbaaccbae4d1c4c2ded37deaa0c7b5b88151bec2022007a5f91f742933be69f6bcc7efb820c435db6c50f611e54e09b8bf97d3e77ff4012103fe138317b7f505764062a2a752e2339afb1ba5435c1de2bf42719dce2c35a3e9ffffffff0240420f00000000001976a9146d0ef49ca42b59112be638bdc751360b1be4995d88ac606a3901000000001976a9141987e0b5410e42dce91cdcea8804cae6b36b7bf588ac00000000
   
   $ trpc getrawtransaction $(trpc getblock `trpc getbestblockhash` 1 | jq -r ".tx[10]")

   020000000143e4a966e927267b74e11aba867bc4356fc1ed0ed98d41633db958ec4b3d7ed6010000006a4730440220671a21fc391ca7d00a3ed01809ffb121c70993670f4c72203514623ad1b985f3022006b7d7a895e6808c6f8165b1944699475628f6dae72edd67bdf9b55da92199fe012103e183b4824996f294db0a04c80fbeacc539b719d09f2fd50d4c8a2e8192db6b74feffffff027ab5cabf3e0000001976a914d361e252335820fe0e1a7991b1a03a57994fa08e88ac0d74cf05000000001976a914210dbd566c705e77802925c6b416aeca13e726b888ac2a121400
   ```

**Comment by sipa** [comment]
Concept ACK

**Comment by laanwj** [comment]
Concept ACK, nice

**Comment by promag** [comment]
Needs rebase.

**Comment by laanwj** [comment]
utACK ec3073a274bf7affe1b8c87a10f75d126f5ac027
verified move-onlyness of 2318affd27de436ddf9d866a4b82eed8ea2e738b
